verilog/analysis/checkers/numeric_format_string_style_rule.cc

```diff
@@ -0,0 +1,141 @@
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(NumericFormatStringStyleRule);
+
+absl::string_view NumericFormatStringStyleRule::Name() { return "numeric-format-string-style"; }
+const char NumericFormatStringStyleRule::kTopic[] = "number-formatting";
+const char NumericFormatStringStyleRule::kMessage[] =
+  "Formatting string must contain proper style-compilant numeric specifiers.";
+
+std::string NumericFormatStringStyleRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat("Checks that string literals with numeric format specifiers "
+    "have proper prefixes for hex and bin values and no prefixes for decimal values.");
+}
+
+void NumericFormatStringStyleRule::HandleToken(const TokenInfo& token) {
+  const auto token_enum = static_cast<verilog_tokentype>(token.token_enum());
+  const absl::string_view text(token.text());
+
+  if (token_enum != TK_StringLiteral)
+    return ;
+
+  for (size_t pos = 0 ; pos < text.size() ;) {
+    const auto& ch = text[pos];
+
+    // Skip ordinary characters
+    if (ch != '%') {
+      ++pos;
+      continue ;
+    }
+
+    // Examine formatting directive
+    for (size_t itr = pos + 1 ; itr < text.size() ; ++itr) {
+      // Skip field with display data size
+      if (absl::ascii_isdigit(text[itr]))
+        continue ;
+
+      const auto radix = text[itr];
+      const auto len = itr - pos + 1;
+
+      // Format radix
+      switch (radix) {
+        // binary value
+        case 'b':
+        case 'B':
+
+        // hexdecimal value
+        case 'h':
+        case 'H':
+        case 'x':
+        case 'X': {
+          // Check for prefix
+          if (pos >= 2 && (text[pos-2] == '0' || text[pos-2] == '\'')) {
+            if (text[pos-1] != absl::ascii_tolower(radix)) {
+              // Report whole prefix with a "0" or "'"
+              violations_.insert(LintViolation(TokenInfo(token.token_enum(),
+                                                         text.substr(pos - 2, len + 2)),
+                                               kMessage));
+            }
+          } else {
+              // Report just radix
+              violations_.insert(LintViolation(TokenInfo(token.token_enum(),
+                                                         text.substr(pos, len)),
+                                               kMessage));
+          }
+          break;
+        }
+
+        // decimal value
+        case 'd':
+        case 'D': {
+          // Detect prefixes starting with "0" and "'"
+          if (pos >= 2 && (text[pos-2] == '0' || text[pos-2] == '\'')) {
+            // Report whole string with a "0" or "'"
+            violations_.insert(LintViolation(TokenInfo(token.token_enum(),
+                                                       text.substr(pos - 2, len + 2)),
+                                             kMessage));
+          }
+          break;
+        }
+
+        // TODO(fangism): Unknown radix, report it somehow?
```

Comment:
Ok to ignore, we can't be 100% certain it is a radix prefix because it is inside a format string as plain text.